nexus/src/app/sled.rs

```diff
@@ -237,4 +239,227 @@ impl super::Nexus {
+    ) -> Result<(), Error> {
+        // For every sled that isn't the sled this instance was allocated to, create
+        // a virtual to physical mapping for each of this instance's NICs.
+        //
+        // For the mappings to be correct, a few invariants must hold:
+        //
+        // - mappings must be set whenever an instance's sled changes (eg.
+        //   during instance creation, migration, stop + start)
+        //
+        // - an instances' sled must not change while its corresponding mappings
+        //   are being created
+        //
+        // - the same mapping creation must be broadcast to all sleds
+        //
+        // A more targeted approach would be to see what other instances share
+        // the VPC this instance is in (or more generally, what instances should
+        // have connectivity to this one), see what sleds those are allocated
+        // to, and only create V2P mappings for those sleds.
+        //
+        // There's additional work with this approach:
+        //
+        // - it means that delete calls are required as well as set calls,
+        //   meaning that now the ordering of those matters (this may also
+        //   necessitate a generation number for V2P mappings)
+        //
+        // - V2P mappings have to be bidirectional in order for both instances's
+        //   packets to make a round trip. This isn't a problem with the
+        //   broadcast approach because one of the sides will exist already, but
+        //   it is something to orchestrate with a more targeted approach.
+        //
+        // TODO-correctness OPTE currently will block instances in different
+        // VPCs from connecting to each other. If it ever stops doing this, the
+        // broadcast approach will create V2P mappings that shouldn't exist.
+
+        let (.., authz_instance, db_instance) =
+            LookupPath::new(&opctx, &self.db_datastore)
+                .instance_id(instance_id)
+                .fetch_for(authz::Action::Read)
+                .await?;
+
+        let instance_nics = self
+            .db_datastore
+            .derive_guest_network_interface_info(&opctx, &authz_instance)
+            .await?;
+
+        // Lookup the physical host IP of the sled hosting this instance
+        let instance_sled_id = db_instance.runtime().sled_id;
+        let physical_host_ip =
+            *self.sled_lookup(&self.opctx_alloc, &instance_sled_id).await?.ip;
+
+        let mut last_sled_id: Option<Uuid> = None;
+        loop {
+            let pagparams = DataPageParams {
+                marker: last_sled_id.as_ref(),
+                direction: dropshot::PaginationOrder::Ascending,
+                limit: std::num::NonZeroU32::new(10).unwrap(),
+            };
+
+            let sleds_page =
+                self.sleds_list(&self.opctx_alloc, &pagparams).await?;
+            let mut join_handles =
+                Vec::with_capacity(sleds_page.len() * instance_nics.len());
```

Comment:
yea, i'd agree with ben here and doing this as as one SQL query. Add follow up issue?

Reply:
opened https://github.com/oxidecomputer/omicron/issues/2770